File: python/hash-table/binary-tree-inorder-traversal[1].py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        res = []
        st = []

        while root or st:
            while root:
                st.append(root)
                root = root.left

            root = st.pop()
            res.append(root.val)
            root = root.right

        return res

File: c/hash-table/binary-tree-inorder-traversal[1].c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* inorderTraversal(struct TreeNode* root, int* returnSize) {
    struct TreeNode *st[1000];
    int top = -1;
    int *arr;
    int idx = 0;
    int count = 32;

    arr = malloc(sizeof(int) * count);
    while(root != NULL || top > -1){
        while(root){
            st[++top] = root;
            root = root->left;
        }

        root = st[top--];
        if(idx == count){
            count <<= 1;
            arr = realloc(arr, sizeof(int) * count);
        }
        arr[idx++] = root->val;
        root = root->right;
    }

    *returnSize = idx;
    return arr;
}
